### DISCUSSION TITLE
How to use state adapt with angular ssr ?

### DISCUSSION THREAD

**Discussion by peterson-umoke** [state-adapt/state-adapt#68] [Q&A]
Hi, @mfp22 and the state adapt team,

I just have to get this off my chest, state adapt is an amazing state management solution, its has even opened up my experience using RxJS, and I fell in love with it so much I just used it to launch a production project.

While trying to pick it up, there were a lot of head scratching moments, some of which were due to me not understanding RxJS and the rest just not enough practical examples on the official documentation.

I just had a request to integrate angular ssr to the platform and I got stuck with a window not defined. If I remove state adapt, everything works fine. I feel the issue might be the redux dev tool, but I am not sure how to go about sort that.

Please how do I resolve that? ... is there any example repo or doc that explains that.

**Comment by mfp22** [comment]
Can you paste the exact error? What version are you using? I may have already fixed this.

I'm really glad to hear that you're enjoying it and it's helping with RxJS as well :)

**Reply by peterson-umoke** [reply]
Here is snippet of the error as shown in my console. It's very vague.

```
4:34:29 AM [vite] Internal server error: window is not defined
      at eval (/<project-name>/node_modules/@state-adapt/angular/fesm2020/state-adapt-angular.mjs:62:29)
      at async instantiateModule (file:///<project-name>/node_modules/@angular-devkit/build-angular/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:56052:9) (x2)
```
state adapt version I am using is ^2.0.5
angular version: ^17.0.0

**Reply by mfp22** [reply]
Okay you can circumvent this by setting it up differently in your project. I fixed this issue in the core libraries, but not in the Angular library. So rather than using `defaultStoreProvider`, you need to define the provider yourself:

```typescript
import { StateAdapt, configureStateAdapt } from '@state-adapt/rxjs';

// ... 
providers: [
//...
{
  provide: StateAdapt,
  useValue: configureStateAdapt(),
},
],
```

Give that a try and let me know if it fixes it.

**Reply by peterson-umoke** [reply]
I did try this, at first it worked, but I when I routed to a lazy loaded component, it resurfaced. here is the error it displays

```
|- ReferenceError: window is not defined
    at eval (/<project-name>/node_modules/@state-adapt/angular/fesm2020/state-adapt-angular.mjs:62:29)
    at async instantiateModule (file:///<project-name>/node_modules/@angular-devkit/build-angular/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:56052:9)

1:37:03 PM [vite] Error when evaluating SSR module /chunk-74CDT44S.mjs: failed to import "/chunk-QNXV2YOI.mjs"
|- ReferenceError: window is not defined
    at eval (/<project-name>/node_modules/@state-adapt/angular/fesm2020/state-adapt-angular.mjs:62:29)
    at async instantiateModule (file:///<project-name>/node_modules/@angular-devkit/build-angular/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:56052:9)
```

I think its the same error as the last time

**Reply by mfp22** [reply]
Are you providing StateAdapt in the lazy module too? I'm not sure why it would do this. It should just use the provider you defined in the root module. `adapt` uses `inject(StateAdapt)`, and there's no `window` reference that I can remember in that code. It's when it's resolving `StateAdapt` that it will look at the provider, and whether that errors just depends on how you defined it.

Either way, I searched the whole StateAdapt codebase and only found the one reference to `window` that will be fixed in the next patch version. 

There's one other issue I need to look at before releasing this.

**Reply by peterson-umoke** [reply]
Sorry for the late reply, but I honestly appreciate your speedy response. Looking forward to the next patch.

**Reply by timsar2** [reply]
Using default nx ssr app config:

**main.ts file:**
`bootstrapApplication(AppComponent, appConfig).catch(err => console.error(err))`

**main.server.ts file:**
`export const config = mergeApplicationConfig(appConfig, serverConfig)`

**app.config.ts file:**
```
export const appConfig: ApplicationConfig = {
  providers: [
    defaultStoreProvider,
    provideClientHydration(
      withHttpTransferCacheOptions({
        includePostRequests: true
      })
    ),
    provideRouter(appRoutes),
    provideAnimations()
  ]
}
```

`export const appRoutes: Route[] = []`
